### PR TITLE
[Merged by Bors] - chore: replace Nat.digits_add_two_add_one with a syntactic matching simp lemma

### PR DESCRIPTION
**Pull request by alexjbest**
This seems more useful once the theory is set up.
Previously
```
Nat.digits 10 (n + 1) = l
```
would simp to
```
(n + 1) % (8 + 2) :: Nat.digits (8 + 2) ((n + 1) / (8 + 2)) = l
```
now it simps to
```
(n + 1) % 10 :: Nat.digits 10 ((n + 1) / 10) = l
```

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by dwrensha** [comment]
This looks good to me.

Before this change, the following fails:
```lean
example : [0,0,0,0,0,0,0,1] = Nat.digits 10 10000000 := by simp
-- unsolved goals
-- ⊢ [0, 0, 0, 0, 0, 0, 1] = Nat.digits (8 + 2) ((9999999 + 1) / (8 + 2))
```

After this change, it succeeds:

```lean
example : [0,0,0,0,0,0,0,1] = Nat.digits 10 10000000 := by simp
--- goals accomplished
```

File: Mathlib/Data/Nat/Digits.lean [hunk 1]
```diff
@@ -115,12 +115,16 @@ theorem digits_one_succ (n : ℕ) : digits 1 (n + 1) = 1 :: digits 1 n :=
...
 #align nat.digits_one_succ Nat.digits_one_succ
 
-@[simp]
 theorem digits_add_two_add_one (b n : ℕ) :
     digits (b + 2) (n + 1) = ((n + 1) % (b + 2)) :: digits (b + 2) ((n + 1) / (b + 2)) := by
   simp [digits, digitsAux_def]
 #align nat.digits_add_two_add_one Nat.digits_add_two_add_one
 
+@[simp]
+lemma digits_of_two_le_of_pos (hm : 2 ≤ m) (hn : 0 < n) :
```
Comment (dwrensha) [inline Mathlib/Data/Nat/Digits.lean:124]: Perhaps `m` should be renamed to `b`, for consistency with the other declarations in this file?

Comment (dwrensha) [reply Mathlib/Data/Nat/Digits.lean:124]: Moreover, currently `m` is an autoimplicit, which I think goes against general mathlib style.

File: Mathlib/Data/Nat/Digits.lean [hunk 2]
```diff
@@ -115,12 +115,16 @@ theorem digits_one_succ (n : ℕ) : digits 1 (n + 1) = 1 :: digits 1 n :=
...
-@[simp]
 theorem digits_add_two_add_one (b n : ℕ) :
     digits (b + 2) (n + 1) = ((n + 1) % (b + 2)) :: digits (b + 2) ((n + 1) / (b + 2)) := by
   simp [digits, digitsAux_def]
 #align nat.digits_add_two_add_one Nat.digits_add_two_add_one
 
+@[simp]
+lemma digits_of_two_le_of_pos {m : ℕ} (hm : 2 ≤ m) (hn : 0 < n) :
+    Nat.digits m n = n % m :: Nat.digits m (n / m) := by
+  rw [Nat.eq_add_of_sub_eq hm rfl, Nat.eq_add_of_sub_eq hn rfl, Nat.digits_add_two_add_one]
```
Comment (alexjbest) [inline Mathlib/Data/Nat/Digits.lean:126]:
```suggestion
lemma digits_of_two_le_of_pos {b : ℕ} (hm : 2 ≤ b) (hn : 0 < n) :
    Nat.digits b n = n % b :: Nat.digits b (n / b) := by
  rw [Nat.eq_add_of_sub_eq hb rfl, Nat.eq_add_of_sub_eq hn rfl, Nat.digits_add_two_add_one]
```

File: Mathlib/Data/Nat/Digits.lean [hunk 3]
```diff
@@ -115,12 +115,16 @@ theorem digits_one_succ (n : ℕ) : digits 1 (n + 1) = 1 :: digits 1 n :=
...
 #align nat.digits_one_succ Nat.digits_one_succ
 
-@[simp]
 theorem digits_add_two_add_one (b n : ℕ) :
     digits (b + 2) (n + 1) = ((n + 1) % (b + 2)) :: digits (b + 2) ((n + 1) / (b + 2)) := by
   simp [digits, digitsAux_def]
 #align nat.digits_add_two_add_one Nat.digits_add_two_add_one
 
+@[simp]
+lemma digits_of_two_le_of_pos {b : ℕ} (hm : 2 ≤ b) (hn : 0 < n) :
```
Comment (alexjbest) [inline Mathlib/Data/Nat/Digits.lean:124]:
```suggestion
lemma digits_of_two_le_of_pos {b : ℕ} (hb : 2 ≤ b) (hn : 0 < n) :
```

**Comment by dwrensha** [comment]
I tried backporting this change to mathlib3, because I as curious as to why things ended up in their current state.

Turns out that, even after making this change there, the equivalent example still fails
```lean
example : [0,0,0,0,0,0,0,1] = nat.digits 10 10000000 := by simp
-- simplify tactic failed to simplify
```

So, it seems to me that the Lean 4 simplifier has a more powerful subgoal discharger that can handle stuff like `2 ≤ 10` and `0 < 1000 / 10`, and this explains why the old `simp` lemma was previously preferred.

**Comment by alexjbest** [comment]
I think simp now uses decide? Maybe that is the reason. Numerals also worked completely differently so I'm not sure even the old one would work on explicit numbers in lean 3 would it?

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**